src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -60,29 +62,164 @@ protected override void Dispose(bool finalize)
+
+                    if (fromPrimaryBranch)
+                    {
+                        // if the solutionChecksum is for primary branch, update primary workspace cache with the solution
+                        return UpdateSolutionIfPossible(solution, workspaceVersion);
+                    }
+
+                    // otherwise, just return the solution
+                    return solution;
+                }
+
+                // we need new solution. bulk sync all asset for the solution first.
+                await assetProvider.SynchronizeSolutionAssetsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                // get new solution info and options
+                var (solutionInfo, options) = await assetProvider.CreateSolutionInfoAndOptionsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                if (fromPrimaryBranch)
+                {
+                    // if the solutionChecksum is for primary branch, update primary workspace cache with new solution
+                    if (TrySetCurrentSolution(solutionInfo, workspaceVersion, options, out var solution))
+                    {
+                        return solution;
+                    }
+                }
+
+                // otherwise, just return new solution
+                var workspace = new TemporaryWorkspace(Services.HostServices, WorkspaceKind.RemoteTemporaryWorkspace, solutionInfo, options);
+                return workspace.CurrentSolution;
             }
+            catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceledAndPropagate(e))
+            {
+                throw ExceptionUtilities.Unreachable;
+            }
+        }
 
+        private Solution? GetAvailableSolution(Checksum solutionChecksum)
+        {
+            var currentSolution = _primaryBranchSolutionWithChecksum;
+            if (currentSolution?.Item1 == solutionChecksum)
+            {
+                // asked about primary solution
+                return currentSolution.Item2;
+            }
+
+            var lastSolution = _lastRequestedSolutionWithChecksum;
+            if (lastSolution?.Item1 == solutionChecksum)
+            {
+                // asked about last solution
+                return lastSolution.Item2;
+            }
+
+            return null;
+        }
+
+        public async Task<Solution> GetSolutionAsync(
+            AssetProvider assetProvider,
+            Checksum solutionChecksum,
+            bool fromPrimaryBranch,
+            int workspaceVersion,
+            CancellationToken cancellationToken)
+        {
+            var availableSolution = GetAvailableSolution(solutionChecksum);
+            if (availableSolution != null)
+            {
+                return availableSolution;
+            }
+
+            // make sure there is always only one that creates a new solution
```

Comment:
i know this is a copy, but i wish comments like thsi explained why.

Reply:
Me too